### PR TITLE
[feat] add trainer factory

### PR DESCRIPTION
**Pull request by geniuspatrick**
Thank you for your contribution to the MindCV repo.
Before submitting this PR, please make sure:

- [x] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindcv/blob/main/CONTRIBUTING.md)
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved terminology
- [ ] You have added unit tests

## Motivation

We abstracted the fragment of the training script about creating a `mindspore.Model` (actually a Trainer) into the function `create_trainer`, a factory method of creating trainers. We believe this abstraction facilitates code readability.

When creating the trainer we need to pass in common components such as networks, optimizer, loss functions etc. as input parameters. Meanwhile, we also need to pass in additional parameters to support Auto Mixed Precision(AMP). Next, we elaborate on the creator's design and principles regarding AMP in details.

1. The level of amp:
   We follow the definition from [MindSpore](https://www.mindspore.cn/docs/zh-CN/master/api_python/amp/mindspore.amp.auto_mixed_precision.html)
   We may subsequently consider **customized** black and white lists

2. The type of loss scale:
   - fixed(w/ or w/o drop_overflow_update)
   - dynamic
   - auto

   For `fixed` or `dynamic`, we will explicitly construct the `LossScaleManager` and pass in `mindspore.Model`. For `auto`, we will not actively construct the `LossScaleManager`, but `mindspore.Model` may use the `LossScaleManager` silently, see [mindspore.train.amp](https://gitee.com/mindspore/mindspore/blob/master/mindspore/python/mindspore/train/amp.py) for details.

3. The value of loss scale
   We will raise an error if the value of loss scale is less than 1. We no longer make a special case for `loss_scale=1`, although for a while we did. You can now set the type of loss scale to `auto` to achieve the same effect.

We have also considered support for customized `TrainStep`. The current customized `TrainStep` supports EMA and gradient clipping. Note: the current customized `TrainStep` can only be used with fixed loss scale without dropping overflow!

## Test Plan

st is already in `tests/tasks/test_train_val_imagenet_subset.py`. Do we need an additional unit test?

## Related Issues and PRs

Nope

### CODE REVIEW

File: mindcv/engine/trainer_factory.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+def create_trainer(
+    network: nn.Cell,
+    loss: nn.Cell,
+    optimizer: nn.Cell,
+    metrics: Union[dict, set],
+    amp_level: str,
+    loss_scale_type: str,
+    loss_scale: float = 1.0,
+    drop_overflow_update: bool = False,
```
Comment (SamitHuang) [inline mindcv/engine/trainer_factory.py:39]: 要不把drop_overflow_update统一放到train one step cell里，这样可以支持任意loss_scaler? 比如
`        if self.drop_overflow_update:
            # if there is no overflow, do optimize
            if not overflow :
                loss = F.depend(loss, self.optimizer(grads))
                if self.ema:
                    self.ema_update()
            else:
                print('Gradient flow-point Overflow. Skip update step.')

        else:
            # still optimizer even overflow
            loss = F.depend(loss, self.optimizer(grads))
            if self.ema:
                self.ema_update()
`

Comment (geniuspatrick) [reply mindcv/engine/trainer_factory.py:39]: 我们自定义的TrainStep类后续要计划要支持所有的loss scale；但是仍应当保留不自定义TrainStep的情况，保持向后兼容性

File: mindcv/engine/trainer_factory.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            # todo: drop_overflow_update. If drop_overflow_update is False, scale_sense should be a number
+            #  instead of cell, and TrainStep should be TrainOneStepCell. If drop_overflow_update is True,
+            #  scale_sense should be FixedLossScaleUpdateCell, and TrainStep should be TrainOneStepWithLossScaleCell.
+            train_step_kwargs["scale_sense"] = nn.FixedLossScaleUpdateCell(loss_scale_value=loss_scale)
+        elif loss_scale_type.lower() == "dynamic":
+            train_step_kwargs["scale_sense"] = nn.DynamicLossScaleUpdateCell(
+                loss_scale_value=loss_scale, scale_factor=2, scale_window=2000
+            )
+        else:
+            raise ValueError(f"Loss scale type only support ['fixed', 'dynamic'], but got{loss_scale_type}.")
```
Comment (Songyuanwei) [inline mindcv/engine/trainer_factory.py:118]: 这的loss scale type和上面的不一样，不支持auto，用户会产生歧义吗？

Comment (geniuspatrick) [reply mindcv/engine/trainer_factory.py:118]: 同样需要自定义的TrainStep支持auto

File: mindcv/engine/trainer_factory.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            ema_decay=ema_decay,
+            clip_grad=clip_grad,
+            clip_value=clip_value,
+        )
+        if loss_scale_type.lower() == "fixed":
+            # todo: drop_overflow_update. If drop_overflow_update is False, scale_sense should be a number
+            #  instead of cell, and TrainStep should be TrainOneStepCell. If drop_overflow_update is True,
+            #  scale_sense should be FixedLossScaleUpdateCell, and TrainStep should be TrainOneStepWithLossScaleCell.
+            train_step_kwargs["scale_sense"] = nn.FixedLossScaleUpdateCell(loss_scale_value=loss_scale)
+        elif loss_scale_type.lower() == "dynamic":
```
Comment (Songyuanwei) [inline mindcv/engine/trainer_factory.py:113]: train_step中目前仅有不检测溢出的情况，如果设置成dynamic loss scale， train_step中应该要检测溢出，但是目前中代码没有检测溢出。

Comment (geniuspatrick) [reply mindcv/engine/trainer_factory.py:121]: 后面加了loss scale的限制
